src/lib/userStorage/UserProfileStorage.js

```diff
@@ -0,0 +1,556 @@
+  setAvatar(avatar: string): Promise<void>;
+  removeAvatar(): Promise<void>;
+  getProfileByWalletAddress(walletAddress: string): Promise<any>;
+  getPublicProfile(key: string, value: string): { [field: string]: string };
+  getProfileFieldValue(field: string): string;
+  getProfileFieldDisplayValue(field: string): string;
+  getProfileField(field: string): ProfileField;
+  getDisplayProfile(): UserModel;
+  getPrivateProfile(): UserModel;
+  getFieldPrivacy(field: string): string;
+  validateProfile(profile: any): Promise<{ isValid: boolean, errors: {} }>;
+  setProfileFieldPrivacy(field: string, privacy: FieldPrivacy): Promise<ACK>;
+  getUserProfile(field?: string): { name: string, avatar: string };
+  deleteProfile(): Promise<boolean>;
+}
+
+//5. implement pubsub for profile changes (one method to subscribe for profile updates, when profile changes notify the subscribers)
+export class UserProfileStorage implements ProfileStorage {
+  profileDefaults: {} = {
+    mobile: '',
+  }
+
+  profileSettings: {} = {
+    fullName: { defaultPrivacy: 'public' },
+    email: { defaultPrivacy: 'private' },
+    mobile: { defaultPrivacy: 'private' },
+    mnemonic: { defaultPrivacy: 'private' },
+    avatar: { defaultPrivacy: 'public' },
+    smallAvatar: { defaultPrivacy: 'public' },
+    walletAddress: { defaultPrivacy: 'public' },
+    username: { defaultPrivacy: 'public' },
+  }
+
+  subscribersProfileUpdates = []
+
+  walletAddressIndex = {}
+
+  //unecrypted profile field values
+  profile: { [key: string]: ProfileField } = {}
+
+  /**
+   * Returns phone with last 4 numbers, and before that ***,
+   * and hide email user characters leaving visible only first and last character
+   * @param {string} fieldType - (Email, mobile or phone) Field name
+   * @param {string} value - Field value
+   * @returns {string} - Returns masked value with *** to hide characters
+   */
+  static maskField = (fieldType: 'email' | 'mobile' | 'phone', value: string): string => {
+    if (fieldType === 'email') {
+      let parts = value.split('@')
+      return `${parts[0][0]}${'*'.repeat(parts[0].length - 2)}${parts[0][parts[0].length - 1]}@${parts[1]}`
+    }
+    if (['mobile', 'phone'].includes(fieldType)) {
+      return `${'*'.repeat(value.length - 4)}${value.slice(-4)}`
+    }
+    return value
+  }
+
+  constructor(wallet: GoodWallet, profiledb: ProfileDB) {
+    // const seed = Uint8Array.from(Buffer.from(pkeySeed, 'hex'))
+    // this.privateKey = TextileCrypto.PrivateKey.fromRawEd25519Seed(seed)
+    this.wallet = wallet
+    this.profiledb = profiledb
+  }
+
+  /**
+   * read raw profile from database
+   */
+  async init() {
+    const rawProfile = await this.profiledb.getProfile()
+    const decryptedProfile = await this._decryptProfileFields(rawProfile)
+    // eslint-disable-next-line array-callback-return
+    Object.keys(this.profileSettings).map(field => {
```

Comment:
Mutating objects isn't a good option in JS
You'd need to use immutable objects arrays as much as it's possible

here you could use mapValues from lodash over the decrypted profile to mirror it onto the new object instance

Reply:
why whats the problem with mutating here? @serdiukov-o-nordwhale

Reply:
here it's not critical but here https://github.com/GoodDollar/GoodDAPP/pull/3358#discussion_r685925614 it could cause issues as we're mutating this.profile